egui/src/widgets/plotting/items.rs

```diff
@@ -0,0 +1,214 @@
+//! Contains items that can be added to a plot.
+
+use std::ops::RangeInclusive;
+
+use super::transform::Bounds;
+use crate::*;
+
+/// A value in the value-space of the plot.
+///
+/// Uses f64 for improved accuracy to enable plotting
+/// large values (e.g. unix time on x axis).
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Value {
+    /// This is often something monotonically increasing, such as time, but doesn't have to be.
+    /// Goes from left to right.
+    pub x: f64,
+    /// Goes from bottom to top (inverse of everything else in egui!).
+    pub y: f64,
+}
+
+impl Value {
+    #[inline(always)]
+    pub fn new(x: impl Into<f64>, y: impl Into<f64>) -> Self {
+        Self {
+            x: x.into(),
+            y: y.into(),
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+/// A horizontal line in a plot, filling the full width
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct HLine {
+    pub(crate) y: f64,
+    pub(crate) stroke: Stroke,
+}
+
+impl HLine {
+    pub fn new(y: impl Into<f64>, stroke: impl Into<Stroke>) -> Self {
+        Self {
+            y: y.into(),
+            stroke: stroke.into(),
+        }
+    }
+}
+
+/// A vertical line in a plot, filling the full width
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct VLine {
+    pub(crate) x: f64,
+    pub(crate) stroke: Stroke,
+}
+
+impl VLine {
+    pub fn new(x: impl Into<f64>, stroke: impl Into<Stroke>) -> Self {
+        Self {
+            x: x.into(),
+            stroke: stroke.into(),
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------
+
+enum Generator {
+    /// Describes a function y = f(x) with an optional range for x and a number of points.
+    Explicit(Box<dyn Fn(f64) -> f64>, Option<RangeInclusive<f64>>, usize),
+    /// Describes a function (x,y) = f(t) with a range for t and a number of points.
+    Parametric(Box<dyn Fn(f64) -> (f64, f64)>, RangeInclusive<f64>, usize),
```

Comment:
```suggestion
    /// Describes a function y = f(x) with a range for x and a number of points.
    Explicit{
        y_from_x: Box<dyn Fn(f64) -> f64>,
        x_range: RangeInclusive<f64>,
        resolution: usize,
    },
    /// Describes a function (x,y) = f(t) with a range for t and a number of points.
    Parametric{
        xy_from_t: Box<dyn Fn(f64) -> (f64, f64)>,
        t_range: RangeInclusive<f64>,
        resolution: usize
    },
```

I see no difference between `None` range and an `-INF..=INF` range, so better to unify to the simpler type